src/hibernate-resume/hibernate-resume-generator.c

```diff
@@ -69,25 +85,84 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
+        const char *offset;
+        dev_t dev;
+        int r;
+
+        r = efi_get_variable_string(EFI_SYSTEMD_VARIABLE(HibernateInfo), &info);
+        if (r == -ENOENT)
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to get EFI variable HibernateInfo: %m");
+
+        info_split = strv_split(info, "=");
+        if (!info_split)
+                return log_oom();
+
+        r = parse_devnum(info_split[0], &dev);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse HibernateInfo device '%s': %m", info_split[0]);
+
+        r = devname_from_devnum(S_IFBLK, dev, &device);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to get resume devname from devnum '%s': %m", info_split[0]);
+
+        offset = info_split[1];
+
+        if (!arg_resume_device) {
+                arg_resume_device = TAKE_PTR(device);
+
+                if (offset) {
+                        arg_resume_offset = strdup(offset);
+                        if (!arg_resume_offset)
+                                return log_oom();
+                }
+        } else {
+                if (!streq(arg_resume_device, device))
+                        log_warning("resume=%s mismatches with HibernateInfo device '%s', proceeding anyway.",
+                                    arg_resume_device, device);
+
+                if (!streq_ptr(arg_resume_offset, offset))
+                        log_warning("resume_offset=%s mismatches with HibernateInfo offset %s, proceeding anyway.",
+                                    strna(arg_resume_offset), strna(offset));
+        }
+
+        return 0;
+}
+#endif
+
 static int process_resume(void) {
-        _cleanup_free_ char *service_unit = NULL, *device_unit = NULL, *lnk = NULL;
+        _cleanup_free_ char *service_unit = NULL, *device_unit = NULL, *link = NULL;
         int r;
 
         if (!arg_resume_device)
                 return 0;
 
-        r = unit_name_from_path_instance("systemd-hibernate-resume", arg_resume_device, ".service",
-                                         &service_unit);
+        r = unit_name_from_path_instance("systemd-hibernate-resume", arg_resume_device, ".service", &service_unit);
         if (r < 0)
                 return log_error_errno(r, "Failed to generate unit name: %m");
 
-        lnk = strjoin(arg_dest, "/" SPECIAL_SYSINIT_TARGET ".wants/", service_unit);
-        if (!lnk)
+        link = path_join(arg_dest, SPECIAL_SYSINIT_TARGET ".wants", service_unit);
+        if (!link)
                 return log_oom();
 
-        (void) mkdir_parents_label(lnk, 0755);
-        if (symlink(SYSTEM_DATA_UNIT_DIR "/systemd-hibernate-resume@.service", lnk) < 0)
-                return log_error_errno(errno, "Failed to create symlink %s: %m", lnk);
+        (void) mkdir_parents_label(link, 0755);
+        if (symlink(SYSTEM_DATA_UNIT_DIR "/systemd-hibernate-resume@.service", link) < 0)
```

Comment:
hmm, so the fact this has been a template unit has always been weird to me, because it is effectively a singleton: you cannot resume from multiple disks at the same time, hence there's no point in having multiple instances. The only reason it was implemented like this to allow easy parameterization via the block device name. But that's out of the window now because of the need to of a secon parameter. Hence I think we should just drop the instantiation. i.e. write out the full unit file, like we do for fsck and similar.